edgelet/hsm-sys/azure-iot-hsm-c/src/edge_hsm_client_x509.c

```diff
@@ -1,72 +1,501 @@
+        {
+            LOG_ERROR("hsm_client_crypto_interface returned NULL");
+            result = NULL;
+        }
+        else
+        {
+            result = interface->hsm_client_crypto_create();
+            if (result != NULL)
+            {
+                g_ref_cnt++;
+            }
+        }
+    }
+
+    return result;
 }
 
-void iothub_x509_hsm_destroy(HSM_CLIENT_HANDLE handle)
+void edge_x509_hsm_destroy(HSM_CLIENT_HANDLE hsm_handle)
 {
-    (void)handle;
-    LOG_ERROR("API unsupported");
+    if (!g_is_x509_initialized)
+    {
+        LOG_ERROR("hsm_client_x509_init not called");
+    }
+    else
+    {
+        if (hsm_handle == NULL)
+        {
+            LOG_ERROR("Null hsm handle parameter");
+        }
+        else if (g_ref_cnt == 0)
+        {
+            LOG_ERROR("Mismatch in overall handle create and destroy calls");
+        }
+        else
+        {
+            const HSM_CLIENT_CRYPTO_INTERFACE* interface = hsm_client_crypto_interface();
+            if (interface == NULL)
+            {
+                LOG_ERROR("hsm_client_crypto_interface returned NULL");
+            }
+            else
+            {
+                interface->hsm_client_crypto_destroy(hsm_handle);
+            }
+            g_ref_cnt--;
+        }
+    }
+}
+
+static CERT_PROPS_HANDLE create_edge_device_properties
+(
+    const char* common_name,
+    uint64_t validity_seconds,
+    const char* issuer_alias
+)
+{
+    CERT_PROPS_HANDLE certificate_props = cert_properties_create();
+    const char* alias = EDGE_DEVICE_ALIAS;
+
+    if (certificate_props == NULL)
+    {
+        LOG_ERROR("Could not create certificate props for %s", alias);
+    }
+    else if (set_common_name(certificate_props, common_name) != 0)
+    {
+        LOG_ERROR("Could not set common name for %s", alias);
+        cert_properties_destroy(certificate_props);
+        certificate_props = NULL;
+    }
+    else if (set_validity_seconds(certificate_props, validity_seconds) != 0)
+    {
+        LOG_ERROR("Could not set validity for %s", alias);
+        cert_properties_destroy(certificate_props);
+        certificate_props = NULL;
+    }
+    else if (set_alias(certificate_props, alias) != 0)
+    {
+        LOG_ERROR("Could not set alias for %s", alias);
+        cert_properties_destroy(certificate_props);
+        certificate_props = NULL;
+    }
+    else if (set_issuer_alias(certificate_props, issuer_alias) != 0)
+    {
+        LOG_ERROR("Could not set issuer alias for %s", alias);
+        cert_properties_destroy(certificate_props);
+        certificate_props = NULL;
+    }
+    else if (set_certificate_type(certificate_props, CERTIFICATE_TYPE_CLIENT) != 0)
+    {
+        LOG_ERROR("Could not set certificate type for %s", alias);
+        cert_properties_destroy(certificate_props);
+        certificate_props = NULL;
+    }
+
+    return certificate_props;
+}
+
+static CERT_INFO_HANDLE create_device_certificate(HSM_CLIENT_HANDLE hsm_handle)
```

Comment:
This is for the case that we only have a signing cert and need to auto create a device cert?

Reply:
Correct.